### PR TITLE
[FLINK-23402][streaming-java][table-planner] Add ShuffleMode option

### PR DESCRIPTION
**Pull request by twalthr**
## What is the purpose of the change

This introduces the config option `execution.shuffle-mode` which is an enum of values `ALL_EXCHANGES_BLOCKING`, `ALL_EXCHANGES_PIPELINED`, and `AUTOMATIC`.

It replaces `table.exec.shuffle-mode` and allows to influence the `GlobalStreamExchangeMode` for both DataStream API and Table API consistently.

## Brief change log

- Use `GlobalStreamExchangeMode.ALL_EDGES_BLOCKING` as new default for batch mode
- Add `ShuffleMode` option
- Refactor `StreamGraphGenerator` to pick up this option
- Adjust table planner to pick up this option and deprecate `table.exec.shuffle-mode`.

## Verifying this change

This change added tests and can be verified as follows: various tests added

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): yes
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? docs

### CODE REVIEW

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/config/ExecutionConfigOptions.java [hunk 1]
```diff
@@ -336,15 +337,20 @@
...
             key("table.exec.shuffle-mode")
                     .stringType()
-                    .defaultValue("ALL_EDGES_BLOCKING")
+                    .noDefaultValue()
                     .withDescription(
                             Description.builder()
                                     .text("Sets exec shuffle mode.")
                                     .linebreak()
+                                    .text(
+                                            "This option is deprecated. Use the core option '%s' instead.",
```
Comment (zentol) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/config/ExecutionConfigOptions.java:351]: Deprecated options are excluded from the documentation, so no one's gonna read this? (maybe put it in the javadocs instead)

File: flink-core/src/main/java/org/apache/flink/api/common/ShuffleMode.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+/**
+ * Mode that defines how data is exchanged between tasks if the shuffling behavior has not been set
+ * explicitly for an individual exchange.
+ *
+ * <p>The shuffle mode depends on the configured {@link ExecutionOptions#RUNTIME_MODE} and is only
+ * relevant for batch executions on bounded streams.
+ *
+ * <p>In streaming mode, upstream and downstream tasks run simultaneously to achieve low latency. An
+ * exchange is always pipelined (i.e. a result record is immediately send to and processed by the
```
Comment (Airblader) [inline flink-core/src/main/java/org/apache/flink/api/common/ShuffleMode.java:35]: send → sent

File: flink-core/src/main/java/org/apache/flink/configuration/ExecutionOptions.java [hunk 3]
```diff
@@ -38,6 +40,32 @@
...
+                                    .text(
+                                            "Mode that defines how data is exchanged between tasks if the shuffling "
+                                                    + "behavior has not been set explicitly for an individual exchange. "
+                                                    + "The shuffle mode depends on the configured '%s' and is only "
+                                                    + "relevant for batch executions on bounded streams.",
+                                            text(RUNTIME_MODE.key()))
+                                    .linebreak()
+                                    .text(
+                                            "In streaming mode, upstream and downstream tasks run simultaneously to achieve low latency. "
+                                                    + "An exchange is always pipelined (i.e. a result record is immediately send to and "
```
Comment (Airblader) [inline flink-core/src/main/java/org/apache/flink/configuration/ExecutionOptions.java:58]: Same here

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphGenerator.java [hunk 4]
```diff
@@ -351,34 +356,67 @@ private void configureStreamGraph(final StreamGraph graph) {
...
+        graph.setStateBackend(stateBackend);
+        graph.setChangelogStateBackendEnabled(changelogStateBackendEnabled);
+        graph.setCheckpointStorage(checkpointStorage);
+        graph.setSavepointDirectory(savepointDir);
+        graph.setGlobalStreamExchangeMode(deriveGlobalStreamExchangeModeStreaming());
+    }
+
+    private GlobalStreamExchangeMode deriveGlobalStreamExchangeModeBatch() {
+        final ShuffleMode shuffleMode = configuration.get(ExecutionOptions.SHUFFLE_MODE);
+        if (shuffleMode == ShuffleMode.ALL_EXCHANGES_PIPELINED) {
```
Comment (Airblader) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphGenerator.java:394]: Why not use switch-case here (and below)? Linting tools / IDEs would then also be able to show warnings/errors if the enum is not cased over exhaustively.

**Review by Airblader — APPROVED** [review]
LGTM.

**Pull request closed without merging**